File: web-admin/src/api/workflow.ts
import { api } from './http';

export interface WorkflowWithDocType {
    id: string;
    workflow_name: string;
    doctype_id: string;
    doctype_name: string;
    is_active: boolean;
    document_status_field: string;
    created_at: string;
    updated_at: string;
}

export interface WorkflowState {
    id?: string;
    workflow_id?: string;
    state_name: string;
    doc_status: number; // 0: Draft, 1: Submitted, 2: Cancelled
    allow_edit_role_id?: string;
    style_variant: string;
    created_at?: string;
}

export interface WorkflowTransitionDetail {
    id: string;
    workflow_id: string;
    state_id: string;
    state_name: string;
    action_name: string;
    next_state_id: string;
    next_state_name: string;
    allowed_role_id: string;
    allowed_role_name: string;
    allowed_role_code: string;
}

export interface WorkflowDetailResponse {
    workflow: WorkflowWithDocType;
    states: WorkflowState[];
    transitions: WorkflowTransitionDetail[];
}

export const workflowApi = {
    listWorkflows: async () => {
        const response = await api.get<WorkflowWithDocType[]>('/workflows');
        return response.data;
    },

    getWorkflowDetail: async (id: string) => {
        const response = await api.get<WorkflowDetailResponse>(`/workflows/${id}`);
        return response.data;
    },

    createWorkflow: async (payload: { workflow_name: string; doctype_id: string; document_status_field?: string }) => {
        const response = await api.post<WorkflowWithDocType>('/workflows', payload);
        return response.data;
    },

    saveWorkflowState: async (workflowId: string, payload: { state_name: string; doc_status: number; allow_edit_role_id?: string; style_variant: string }) => {
        const response = await api.post<WorkflowState>(`/workflows/${workflowId}/states`, payload);
        return response.data;
    },

    saveWorkflowTransition: async (workflowId: string, payload: { state_id: string; action_name: string; next_state_id: string; allowed_role_id: string }) => {
        const response = await api.post<WorkflowTransitionDetail>(`/workflows/${workflowId}/transitions`, payload);
        return response.data;
    },

    deleteWorkflow: async (id: string) => {
        const response = await api.delete(`/workflows/${id}`);
        return response.data;
    },

    applyWorkflowAction: async (payload: {
        workflow_id: string;
        document_id: string;
        current_state_name: string;
        action_name: string;
        comments?: string;
    }) => {
        const response = await api.post('/workflows/apply-action', payload);
        return response.data;
    },
};

File: crates/core/src/domain/entities/workflow.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Workflow {
    pub id: Uuid,
    pub workflow_name: String,
    pub doctype_id: Uuid,
    pub is_active: bool,
    pub document_status_field: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct WorkflowState {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub state_name: String,
    pub doc_status: i32,
    pub allow_edit_role_id: Option<Uuid>,
    pub style_variant: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct WorkflowTransition {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub state_id: Uuid,
    pub action_name: String,
    pub next_state_id: Uuid,
    pub allowed_role_id: Uuid,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct WorkflowActionLog {
    pub id: Uuid,
    pub workflow_id: Uuid,
    pub document_id: Uuid,
    pub action_by_user_id: Uuid,
    pub from_state: String,
    pub action_name: String,
    pub to_state: String,
    pub comments: Option<String>,
    pub created_at: DateTime<Utc>,
}
